include/lsst/partition/ConfigStore.h

```diff
@@ -0,0 +1,290 @@
+     * 
+     * It's allowed to call this method many times for the same or different objects.
+     * Parameters read from the input object will be merged into the store.
+     * Empty objects are allowed on the input.
+     *
+     * @param config  A JSON object to be merged into the store.
+     * @throw std::invalid_argument  If the input value is not a JSON Object.
+     */
+    void add(nlohmann::json const& config);
+
+    /**
+     * Merge values of the command-line parameters into the store.
+     *
+     * @note Interpret empty parameters as boolean flags set as 'true'. Also ignore parameters for which
+     *   only their default values are available (since they were not present in the command line),
+     *   unless the parameters aren't present in the store. In this approach, the last explicitly
+     *   set value would always have the higher priority.
+     *
+     * @param vm  A collection of parameters to be merged.
+     * @throw ConfigStore  For unsupported types of the input parameters.
+     */
+    void add(boost::program_options::variables_map const& vm);
+
+    /**
+     * Set/update a value of a parameter at the specified path.
+     *
+     * @param path  A path to the parameter.
+     * @param value A value of the parameter.
+     * @throw std::invalid_argument The path is not valid.
+     */
+    template <typename T>
+    void set(std::string const& path, T const& value) { _config[_path2pointer(path)] = value; }
+
+    /**
+     * Extract a value of an existing parameter given its expected type.
+     *
+     * @param path  A path to a parameter.
+     * @return A value of requested given type.
+     * @throw std::invalid_argument The path is not valid.
+     * @throw ConfigStore  If using incorrect types in type conversion.
+     */
+    template <typename T>
+    T get(std::string const& path) const { return detail::Value<T>::convert(_get(path), path); }
+
+    /**
+     * Check if the specified parameter exists in the configuration.
+     *
+     * @param path  A path to a parameter.
+     * @return 'true' if the specified parameter exists at the specified path.
+     * @throw std::invalid_argument The path is not valid.
+     */
+    bool has(std::string const& path) const;
+
+    /**
+     * This method should be used instead of 'has(path)' or 'get<bool>(path)' for checking
+     * a status of parameters meant to be used as flags. A reason why this method is needed is
+     * because JSON allows attributes of the 'bool' type which could be explicitly set to 'true'
+     * or 'false'. In this case just relying on 'has(path)' may produce incorrect results.
+     * Method 'has(path)' would only work for parameters found in the command line.
+     * On the other hand, relying on 'get<bool>(path)' would only work for the parameters
+     * which were explicitly set from JSON. Using the getter method would result in an exception
+     * if the parameter wasn't explicitly specified in the JSON file.
+     *
```

Comment:
Can you come up with a more elegant solution where the caller doesn't have to pick the correct function based on expected type? As is, this is error prone.

Reply:
I hear two questions here:

1) If it would be better to rely on some form of the method overloading based on a return type.

I know some libraries do that, including `nlohmann::json`. Personally, I don't find this safe as this solution has many corner cases and unpleasant surprises. To me, passing a desired type explicitly to a parameter's value retreival method `T get<T>(name)` seems like the best solution.

Besides, when I was migratng the package to the new `ConfigStore` API I was trying to preserve the original code which uses type specifications.

2) If having method `bool flag(name)` is the right approach compared to some "more elegant solution".

The answer is found in the comment to the method. The original code was relying on "counting" of the number of command line parameters of the given name. This doesn't work with `JSON` where a boolean parameter a) may be present or absent, or b) it may be set to `true` or `false`. So, I decided the safest approach would be to express user's intent as a request for a presence of some "flag" which could be `true` or `false`, regardless if it was found in the command line, or if it was (or was ot) found in ` `JSON` file. To me a concept of the "flag" is close to the actual intent of the client's code than someting like `bool get<bool>(name)`.